### DISCUSSION TITLE
Question: How to update Docker LXC

### DISCUSSION THREAD

**Discussion by Kirbo** [tteck/Proxmox#485] [Help Needed] [answered]
Hi,
I've installed the [Docker LXC](https://github.com/tteck/Proxmox/blob/main/ct/docker-v4.sh) and now my Portainer tells me there's a new version available.

<img width="260" alt="image" src="https://user-images.githubusercontent.com/1077974/188597298-c6f5fbe3-1d86-459b-b9d0-8656c679e3bd.png">

I did try updating via `apt`, by using the suggested command
<img width="344" alt="image" src="https://user-images.githubusercontent.com/1077974/188597399-a1c807f5-f085-488f-b503-d9de7a2c03f4.png">
...as well as using the one I use on my other debians:
```
apt update ; apt upgrade ; apt-get clean ; apt-get autoremove
```
...but it only updated containerd and some other package.

Then I tried pulling the latest image
```
docker pull portainer/portainer-ce:latest
```
...and it did actually pull the latest image, but even though I restarted the whole VM, I'm still getting the notification in Portainer that there's a version 2.15.0 available.

So, how should I update the Portainer?

I noticed that on`Home Assistant Container LXC` using the `./update` and there choosing the `8) Just Update Containers` it did update the Portainer on that VM.

**Comment by tteck** [comment, marked as answer]
Run in the Docker LXC console
```
docker stop portainer
```
```
docker rm portainer
```
```
docker run -d -p 8000:8000 -p 9000:9000 --name=portainer --restart=always -v /var/run/docker.sock:/var/run/docker.sock -v portainer_data:/data portainer/portainer-ce:latest
```

**Reply by up2you1** [reply]
Sorry to ask, should this work for the Alpine docker lxc too? As I’ve just tried it. Seems to have accepted the commands but doesn’t actually update Portainer. 

Thanks.

**Reply by johndoe0815** [reply]
Any feedback on this?

**Reply by johndoe0815** [reply]
Got it. What is missing is pulling the latest portainer image.
So after rm portainer, you need to execute
`docker pull portainer/portainer-ce:latest`
Only then does the docker run command give you the newest version.

**Reply by up2you1** [reply]
Awesome! Many thanks! Worked perfectly

**Comment by Kirbo** [comment]
Thanks, this seemed to do the trick! 🙂

**Comment by shinyone** [comment]
When I tried the answer there was an issue with the ports from the given answer in this thread.

From [Docker LXC](https://tteck.github.io/Proxmox/)
> Portainer Interface: (https) IP:9443

For simplicity I created a script for all commands for the next time.

1) create a shell script.  i.e.  
```
nano updatePortainer.sh
```
2) add commands from this thread.   Note change on port 9443
```
#! /bin/bash

docker stop portainer
docker rm portainer
docker pull portainer/portainer-ce:latest
docker run -d -p 8000:8000 -p 9443:9443 --name=portainer --restart=always -v /var/run/docker.sock:/var/run/docker.sock -v portainer_data:/data portainer/portainer-ce:latest
```
3) add execute permissions
```
chmod u+x updatePortainer.sh
``` 
4) run script (to upgrade)
```
./updatePortainer.sh
```

**Reply by johndoe0815** [reply]
When trying to update on LXC again, I do now receive the following error message after the last docker run command:

> 3a3cb70129f97990c4b949162d30ee86706474191c3638745094f537da489951
> docker: Error response from daemon: failed to create task for container: failed to create shim task: OCI runtime create failed: runc create failed: unable to start container process: error during container init: error running hook #0: error running hook: exit status 127, stdout: , stderr: Error loading shared library libdevmapper.so.1.02: No such file or directory (needed by libnetwork-setkey)
> Error relocating libnetwork-setkey: dm_task_set_add_node: symbol not found
> Error relocating libnetwork-setkey: dm_udev_get_sync_support: symbol not found
> Error relocating libnetwork-setkey: dm_task_add_target: symbol not found
> Error relocating libnetwork-setkey: dm_task_create: symbol not found
> Error relocating libnetwork-setkey: dm_task_set_cookie: symbol not found
> Error relocating libnetwork-setkey: dm_task_deferred_remove: symbol not found
> Error relocating libnetwork-setkey: dm_cookie_supported: symbol not found
> Error relocating libnetwork-setkey: dm_udev_set_sync_support: symbol not found
> Error relocating libnetwork-setkey: dm_get_library_version: symbol not found
> Error relocating libnetwork-setkey: dm_log_with_errno_init: symbol not found
> Error relocating libnetwork-setkey: dm_task_get_info: symbol not found
> Error relocating libnetwork-setkey: dm_get_next_target: symbol not found
> Error relocating libnetwork-setkey: dm_udev_wait: symbol not found
> Error relocating libnetwork-setkey: dm_set_dev_dir: symbol not found
> Error relocating libnetwork-setkey: dm_task_set_sector: symbol not found
> Error relocating libnetwork-setkey: dm_task_set_name: symbol not found
> Error relocating libnetwork-setkey: dm_task_destroy: symbol not found
> Error relocating libnetwork-setkey: dm_task_run: symbol not found
> Error relocating libnetwork-setkey: dm_task_set_message: symbol not found
> Error relocating libnetwork-setkey: dm_task_get_deps: symbol not found
> Error relocating libnetwork-setkey: dm_task_get_driver_version: symbol not found: unknown.

The previous pull command worked fine, so no clue where this is coming from. Anyone having any idea what needs to be done to fix this?

**Reply by johndoe0815** [reply]
Answering to my own issue:
I had also updated docker before.
In another try, after updating docker, I first rebooted the LXC and then did the portainer update with the same commands as before. This time it worked without any issues.

**Reply by JacobMillward** [reply]
Just confirming that I had the same problem as @johndoe0815, and rebooting the LXC sorted it out
